### ISSUE TITLE
online load: reading data of "not null" columns in Spark causes data load failed

### ISSUE THREAD

**Issue by vagetablechicken** [4paradigm/OpenMLDB#1271]
related to #1170 
Why the error is "write null to non-null column `device_id` failed"? Not the column `std_time`?
**We can set any nullable column to `null`, it will got the same error.** code version ef2e6e826
e.g.
```
device_id,event_type,ability_id,category_first_id,std_time,std_date,testno
aaa,a,a,100,2022-02-07 20:54:20,2022-02-01,null
```

We should check about spark read df. `openmldb-spark-connector` has ut about insert null, it's ok, so the source df may have some error.

**Comment by vagetablechicken** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [improvement for CSV import](https://trello.com/c/eSTnTgIY/13-improvement-for-csv-import)

**Comment by vagetablechicken** [comment]
When we use `DataFrameReader` read csv with `schema()`, the dataframe logicalplan.schema.fields won't get nullable field.
I'll fix it.

**Comment by vagetablechicken** [comment]
CSVFileFormat non-streaming file based datasource
When creating HadoopFsRelation in DataSource.scala, schema will be converted to all nullable.
Can't find a config to change it, so we need to reset the schema after read.

**Comment by dl239** [comment]
#1341 merged to branch-0.4
